iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt

```diff
@@ -0,0 +1,129 @@
+    private val reactiveRedisTemplate: ReactiveRedisTemplate<String, ByteArray>,
+    private val mapper: ObjectMapper,
+) : DeviceRepository by mongoDeviceRepository {
+
+    override fun save(device: MongoDevice): Mono<MongoDevice> {
+        return mongoDeviceRepository.save(device).doOnSuccess {
+            saveDeviceWithRetries(it)
+        }
+    }
+
+    override fun findById(deviceId: String): Mono<MongoDevice> {
+        val key = createDeviceKey(deviceId)
+        return reactiveRedisTemplate.opsForValue().get(key)
+            .handle { item, sink ->
+                if (item.isEmpty()) {
+                    sink.error(EntityNotFoundException("Device with id $deviceId not found"))
+                } else {
+                    runCatching { mapper.readValue<MongoDevice>(item) }
+                        .onSuccess { sink.next(it) }
+                        .onFailure {
+                            reactiveRedisTemplate.unlink(key)
+                            sink.error(it)
+                        }
+                }
+            }
+            .switchIfEmpty { findInMongoAndWriteToRedis(deviceId) }
+            .onErrorResume(::isErrorFromRedis) {
+                mongoDeviceRepository.findById(deviceId)
+            }
+    }
+
+    override fun deleteById(deviceId: String): Mono<Unit> {
+        val key = createDeviceKey(deviceId)
+        return mongoDeviceRepository.deleteById(deviceId)
+            .flatMap {
+                reactiveRedisTemplate.unlink(key)
+                    .onErrorResume(::isErrorFromRedis) {
+                        log.warn("Redis failed to remove user from cache", it)
+                        Mono.empty()
+                    }.thenReturn(Unit)
+            }.thenReturn(Unit)
+    }
+
+    private fun saveDeviceToRedis(device: MongoDevice): Mono<MongoDevice> {
+        val key = createDeviceKey(device.id.toString())
+        val byteArray = mapper.writeValueAsBytes(device)
+        return reactiveRedisTemplate.opsForValue().set(
+            key,
+            byteArray,
+            Duration.ofMinutes(redisExpirationTimeoutInMinutes)
+        ).thenReturn(device)
+    }
+
+    private fun saveDeviceWithRetries(device: MongoDevice) {
+        saveDeviceToRedis(device).retryWhen(
+            Retry.backoff(MAX_RETRIES, Duration.ofSeconds(RETRY_BACKOFF_SECONDS))
+                .filter { error -> isErrorFromRedis(error) }
+                .doBeforeRetry { retrySignal ->
+                    log.warn(
+                        "Retrying saving device to Redis because of {}:",
+                        retrySignal.failure().message
+                    )
+                }
+        )
+            .subscribeOn(Schedulers.boundedElastic())
+            .subscribe()
+    }
+
+    private fun findInMongoAndWriteToRedis(deviceId: String): Mono<MongoDevice> {
+        val key = createDeviceKey(deviceId)
+        return mongoDeviceRepository.findById(deviceId).flatMap {
```

Comment:
I see that sometimes a reactor operator is located on the same line, and sometimes it's placed on the new line

be consistent